xdsl/dialects/arith.py

```diff
@@ -40,12 +40,23 @@ def from_int_and_width(val: int | IntAttr,
             result_types=[typ],
             attributes={"value": IntegerAttr.from_params(val, typ)})
 
-    # To add tests for this constructor
     @staticmethod
     def from_float_and_width(val: float | FloatAttr[_FloatTypeT],
-                             typ: _FloatTypeT) -> Constant:
+                             typ: int | _FloatTypeT) -> Constant:
```

Comment:
I feel like the logic got a bit more confusing. It might be worth reworking it a bit, and adding a comment of the behaviour.
```
if isinstance(typ, int):
  typ = FloatAttr.type_for_width(typ) # would need to add this method

if isinstance(val, FloatAttr):
  if val.typ != typ:
    val = FloatAttr(val.value, typ)
else:
  val = FloatAttr(val, typ)
```

Reply:
Although your idea of the error is better than a silent cast from one `typ` to another